test/cpp/api/tensor.cpp

```diff
@@ -194,6 +196,57 @@ TEST(TensorTest, ContainsCorrectValuesForManyValuesVariable) {
   ASSERT_TRUE(almost_equal(tensor[2], 3.125));
 }
 
+TEST(TensorTest, ContainsCorrectValuesForMultidimValuesVariable) {
+  {
+    auto tensor = torch::tensor({{1, 2}, {3, 4}});
```

Comment:
(doesn't need to be addressed in this PR) In python, one can do something like `x = torch.tensor([])`. Does that work here?

Also, can you create a one element tensor with `torch::tensor(3.14)`?

Reply:
`torch::tensor({})` doesn't work at the moment because we would need to solve the ambiguous overload problem:
```
/data/libtorch/test/example-app.cpp:6:32: error: call of overloaded ‘tensor(<brace-enclosed initializer list>)’ is ambiguous
   std::cout << torch::tensor({}) << std::endl;
                                ^
In file included from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/types.h:7:0,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/data/dataloader_options.h:4,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/data/dataloader/base.h:3,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/data/dataloader/stateful.h:3,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/data/dataloader.h:3,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/data.h:3,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/all.h:4,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/torch.h:3,
                 from /data/libtorch/test/example-app.cpp:1:
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:644: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<unsigned char>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:766: note: candidate: at::Tensor torch::tensor(std::initializer_list<unsigned char>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:887: note: candidate: at::Tensor torch::tensor(uint8_t)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:1603: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<signed char>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:1724: note: candidate: at::Tensor torch::tensor(std::initializer_list<signed char>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:1843: note: candidate: at::Tensor torch::tensor(int8_t)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:2562: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<short int>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:2685: note: candidate: at::Tensor torch::tensor(std::initializer_list<short int>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:2806: note: candidate: at::Tensor torch::tensor(int16_t)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:3507: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<int>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:3624: note: candidate: at::Tensor torch::tensor(std::initializer_list<int>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:3737: note: candidate: at::Tensor torch::tensor(int)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:4450: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<long int>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:4572: note: candidate: at::Tensor torch::tensor(std::initializer_list<long int>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:4693: note: candidate: at::Tensor torch::tensor(int64_t)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:5404: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<float>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:5525: note: candidate: at::Tensor torch::tensor(std::initializer_list<float>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:5642: note: candidate: at::Tensor torch::tensor(float)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:6354: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<double>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:6477: note: candidate: at::Tensor torch::tensor(std::initializer_list<double>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:6596: note: candidate: at::Tensor torch::tensor(double)
```
If the user wants to create an empty tensor, they can use `torch::randn({0})` for now. And if they really want `torch::tensor({})` we can try to find ways to fix the overload problem.

Reply:
and yup `torch::tensor(3.14)` already works: https://github.com/pytorch/pytorch/blob/b63f8ef2c912ddfad645a6ca15fdabd871a2f37a/test/cpp/api/tensor.cpp#L284

Reply:
Yeah, it's fine to not support it in this iteration. `torch::randn({0})` (and `torch.randn([0])) gives a tensor of size `(0,)` right? This is different from `torch.randn([])`, which gives a tensor of size `()`.

Reply:
Yes I think we have the following equivalence:
- `torch::randn({0})` is the same as `torch.randn(0)`, which gives a tensor of size `(0,)` and numel `0`.
- `torch::randn({})` is the same as `torch.randn([])`, which gives a tensor of size `()` and numel `1`.